pilot/pkg/networking/core/v1alpha3/extension/wasmplugin.go

```diff
@@ -0,0 +1,169 @@
+}
+
+// AddWasmPluginsToMutableObjects adds WasmPlugins to HTTP filterChains
+// Note that the slices in the map must already be ordered by plugin
+// priority! This will be the case for maps returned by PushContext.WasmPlugin()
+func AddWasmPluginsToMutableObjects(
+	mutable *networking.MutableObjects,
+	extensionsMap map[extensions.PluginPhase][]*model.WasmPluginWrapper,
+) {
+	if mutable == nil {
+		return
+	}
+
+	for fcIndex, fc := range mutable.FilterChains {
+		// we currently only support HTTP
+		if fc.ListenerProtocol != networking.ListenerProtocolHTTP {
+			continue
+		}
+		mutable.FilterChains[fcIndex].HTTP = injectExtensions(fc.HTTP, extensionsMap)
+	}
+}
+
+func injectExtensions(filterChain []*hcm_filter.HttpFilter, exts map[extensions.PluginPhase][]*model.WasmPluginWrapper) []*hcm_filter.HttpFilter {
+	// copy map as we'll manipulate it in the loop
+	extMap := make(map[extensions.PluginPhase][]*model.WasmPluginWrapper)
+	for phase, list := range exts {
+		extMap[phase] = []*model.WasmPluginWrapper{}
+		extMap[phase] = append(extMap[phase], list...)
+	}
+	newHTTPFilters := make([]*hcm_filter.HttpFilter, 0)
+	// The following algorithm tries to make as few assumptions as possible about the filter
+	// chain - it might contain any number of filters that will have to retain their ordering.
+	// The one assumption we make is about the ordering of the builtin filters. This is used to
+	// position WasmPlugins relatively to the builtin filters according to their phase: when
+	// we see the Stats filter, we know that all WasmPlugins with phases AUTHN, AUTHZ and STATS
+	// must be injected before it. This method allows us to inject WasmPlugins in the right spots
+	// while retaining any filters that were unknown at the time of writing this algorithm,
+	// in linear time. The assumed ordering of builtin filters is:
+	//
+	// 1. Istio JWT, 2. Istio AuthN, 3. RBAC, 4. Stats, 5. Metadata Exchange
+	//
+	// The only exception being ext-auth, where two additional filters are at the beginning:
+	//
+	// -1. RBAC, 0. ext-auth
+	//
+	// In this case, we have to just skip over the RBAC filter instead of using it for phase
+	// positioning. WasmPlugins with phase `AUTHZ` will still be injected before the 'real'
+	// RBAC filter at position 3. The ext-auth filter will be treated like any unknown filter
+	// and simply skipped while retaining its position in the chain.
+	for index, httpFilter := range filterChain {
+		switch httpFilter.Name {
+		case securitymodel.EnvoyJwtFilterName:
+			newHTTPFilters = popAppend(newHTTPFilters, extMap, extensions.PluginPhase_AUTHN)
+			newHTTPFilters = append(newHTTPFilters, httpFilter)
+		case securitymodel.AuthnFilterName:
+			newHTTPFilters = popAppend(newHTTPFilters, extMap, extensions.PluginPhase_AUTHN)
+			newHTTPFilters = append(newHTTPFilters, httpFilter)
+		case authzmodel.RBACHTTPFilterName:
+			// when using ext-auth, the first filter will be an RBAC filter. we ignore that one
+			if index > 0 {
```

Comment:
What if the user only has authz (non ext-authz)? In this case the first filter will be the RBAC filter for the ALLOW/DENY action (non ext-authz). 

I think it's not reliable to assume it's an RBAC filter for ext-authz by just checking the index of the filter, this might not be true for some configurations and hard to maintain in the long term, it would be better to use some explicit fields to pass this information.

Reply:
you're right; just looking at the index won't work. This will need some more thought. I'll remove the ext-authz support for now, it'll have to come later

Reply:
we can skip the support of ext-authz but should not break when ext-authz is used. The current code actually does not work when ext-authz is used (it would take the very first RBAC filter as AUTHZ position)

a quick way is probably to search the filter chain backwards (from end to beginning):
1. If the current filter is neither RBAC nor ext-authz filter, continue the search
2. If the current filter is ext-authz filter, this means there is no authorization policy of action ALLOW/DENY/AUDIT, stop the loop and treat it as no authz found
3. If the current filter is RBAC filter, you can treat it as the authz policy of action ALLOW/DENY/AUDIT and inject the WASM filter accordingly

This will not support ext-authz filter but also won't break when ext-authz is being used.